base/linalg/generic.jl

```diff
@@ -688,9 +688,18 @@ julia> dot([1; 1], [2; 3])
 
 julia> dot([im; im], [1; 1])
 0 - 2im
+
+julia> dot([[0 1; -1 0], [im 0; 0 im]], [1; 1])
+2×2 Array{Complex{Int64},2}:
+ 0-1im  -1+0im
+ 1+0im   0-1im
 ```
 """
 dot(x::AbstractVector, y::AbstractVector) = vecdot(x, y)
+dot(x::AbstractArray, y::Number) = ctranspose(x) * y
+dot(x::Number, y::AbstractArray) = conj(x) * y
+dot(x::AbstractArray, y::AbstractArray) = ctranspose(x) * y
```

Comment:
the `dot(A,B)` call which currently returns a number should really be called `inner(A,B)`, see my suggestion [here](https://github.com/JuliaLang/julia/pull/22220#issuecomment-306185114)